src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        protected abstract class Analyzer<TSource, TDestination> : AnalyzerBase<TSource, TDestination>
+            where TSource : SyntaxNode
+            where TDestination : SyntaxNode
+        {
+            public Analyzer(SemanticModel semanticModel, CancellationToken cancellationToken) : base(semanticModel, cancellationToken) { }
+
+            protected ImmutableArray<IAnonymousFunctionOperation> FindAnonymousFunctionsFromParentInvocationOperation(SyntaxNode node)
+            {
+                var operation = FindParentInvocationOperation(node);
+
+                if (operation != null)
+                {
+                    return ImmutableArray
+                        .CreateRange(operation.Arguments.Where(a => a.Value.Kind == OperationKind.DelegateCreation)
+                        .Select(argumentOperation => (argumentOperation.Value as IDelegateCreationOperation).Target as IAnonymousFunctionOperation));
+                }
+                else
+                {
+                    return ImmutableArray.Create<IAnonymousFunctionOperation>();
+                }
+            }
+
+            private IInvocationOperation FindParentInvocationOperation(SyntaxNode node)
+            {
+                var operation = GetOperation(node)?.Parent;
+                while (operation?.Kind != OperationKind.Invocation)
+                {
+                    operation = operation?.Parent;
+                    if (operation is null)
+                    {
+                        return null;
+                    }
+                }
+
+                return operation as IInvocationOperation;
+            }
+
+            protected IOperation GetOperation(SyntaxNode node)
+            {
+                return _semanticModel.GetOperation(node, _cancellationToken);
+            }
+
+            protected ImmutableArray<string> GetIdentifierNames(SemanticModel semanticModel, SyntaxNode node, CancellationToken cancellationToken)
+            {
+                var operation = GetOperation(node);
+                if (operation == null)
+                {
+                    return default;
+                }
+
+                var builder = ImmutableArray.CreateBuilder<string>();
+
+                while (operation is IPropertyReferenceOperation)
```

Comment:
>IPropertyReferenceOperatio [](start = 36, length = 26)

```operation is IPropertyReferenceOperatio propertyReference```? To avoid another cast inside the loop